pkg/whiteflag/white_flag.go

```diff
@@ -0,0 +1,282 @@
+		(branchVisited || tangle.SolidEntryPointsContain(headTxBranchHash) || branchConfirmed) {
+
+		// if the bundle is conflicting or a value spam bundle,
+		// we don't incorporate it as part of the mutations
+		bundle := cachedBundle.GetBundle()
+
+		visited[string(currentTx.GetTxHash())] = struct{}{}
+		stack.Remove(ele)
+
+		// exclude zero or spam value bundles
+		mutations := bundle.GetLedgerChanges()
+		if bundle.IsValueSpam() || len(mutations) == 0 {
+			wfConf.TailsExcludedZeroValue = append(wfConf.TailsExcludedZeroValue, currentTx.GetTxHash())
+			return nil
+		}
+
+		var conflicting bool
+
+		// contains the updated mutations from this bundle against the
+		// current mutations of the milestone's confirming cone (or previous ledger state).
+		// we only apply it to the milestone's confirming cone mutations if
+		// the bundle doesn't create any conflict.
+		patchedState := make(map[string]int64)
+		validMutations := make(map[string]int64)
+
+		for addr, change := range mutations {
+
+			// load state from milestone cone mutation or previous milestone
+			balance, has := wfConf.NewAddressState[addr]
+			if !has {
+				balanceStateFromPreviousMilestone, _, err := tangle.GetBalanceForAddressWithoutLocking(hornet.Hash(addr))
+				if err != nil {
+					return fmt.Errorf("%w: unable to retrieve balance of address %s", err, addr)
+				}
+				balance = int64(balanceStateFromPreviousMilestone)
+			}
+
+			// note that there's no overflow of int64 values here
+			// as a valid bundle's transaction can not spend more than the total supply,
+			// meaning that newBalance could be max 2*total_supply or min -total_supply.
+			newBalance := balance + change
+
+			// on below zero or above total supply the mutation is invalid
+			if newBalance < 0 || math.AbsInt64(newBalance) > consts.TotalSupply {
+				conflicting = true
+				break
+			}
+
+			patchedState[addr] = newBalance
+			validMutations[addr] = validMutations[addr] + change
+		}
+
+		if conflicting {
+			wfConf.TailsExcludedConflicting = append(wfConf.TailsExcludedConflicting, currentTx.GetTxHash())
+			return nil
+		}
+
+		// mark the given tail to be part of milestone ledger changing tail inclusion set
+		wfConf.TailsIncluded = append(wfConf.TailsIncluded, currentTx.GetTxHash())
+
+		// incorporate the mutations in accordance with the previous mutations
+		// in the milestone's confirming cone/previous ledger state.
+		for addr, balance := range patchedState {
+			wfConf.NewAddressState[addr] = balance
+		}
+
+		// incorporate the mutations in accordance with the previous mutations
+		for addr, mutation := range validMutations {
+			wfConf.AddressMutations[addr] = mutation
+		}
+
+		return nil
+	}
+
+	if !tangle.SolidEntryPointsContain(headTxTrunkHash) && !trunkVisited && !trunkConfirmed {
+		stack.PushFront(headTxTrunkHash)
+		return nil
+	}
+
+	if !tangle.SolidEntryPointsContain(headTxBranchHash) && !branchVisited && !branchConfirmed {
+		stack.PushFront(headTxBranchHash)
```

Comment:
In case `bytes.Equal(headTxTrunkHash, headTxBranchHash)`, wouldn't we push the same transaction two times to the stack?